addon/utils/deprecate-subclassing.js

```diff
@@ -0,0 +1,24 @@
+import { DEBUG } from '@glimmer/env';
+import { deprecate } from '@ember/debug';
+
+export default function deprecateSubclassing(target) {
+  if (DEBUG) {
+    const originalInit = target.prototype.init;
+
+    if (originalInit) {
+      target.prototype.init = function () {
+        deprecate(
```

Comment:
Note that I used `deprecate()` here instead of `warn()`, although this is strictly not a real deprecation. But this allows us to test it easily, and might also be easier to process, as the deprecation message are aggregated in the Ember inspector, and can be silenced if needed. Does that make sense?